rinja_parser/src/node.rs

```diff
@@ -1203,97 +1203,130 @@ impl<'a> Comment<'a> {
-        *b"self____",
-        *b"Self____",
-        *b"true____",
-        *b"type____",
+    const KW4: &[[u8; MAX_REPL_LEN]] = &[
+        *b"r#else____",
+        *b"r#enum____",
+        *b"r#impl____",
+        *b"r#move____",
+        *b"r#priv____",
+        *b"r#true____",
+        *b"r#type____",
     ];
-    const KW5: &[[u8; MAX_KW_LEN]] = &[
-        *b"async___",
-        *b"await___",
-        *b"break___",
-        *b"const___",
-        *b"crate___",
-        *b"false___",
-        *b"final___",
-        *b"macro___",
-        *b"match___",
-        *b"super___",
-        *b"trait___",
-        *b"union___",
-        *b"where___",
-        *b"while___",
-        *b"yield___",
+    const KW5: &[[u8; MAX_REPL_LEN]] = &[
+        *b"r#async___",
+        *b"r#await___",
+        *b"r#break___",
+        *b"r#const___",
+        *b"r#crate___",
+        *b"r#false___",
+        *b"r#final___",
+        *b"r#macro___",
+        *b"r#match___",
+        *b"r#trait___",
+        *b"r#where___",
+        *b"r#while___",
+        *b"r#yield___",
     ];
-    const KW6: &[[u8; MAX_KW_LEN]] = &[
-        *b"become__",
-        *b"extern__",
-        *b"return__",
-        *b"static__",
-        *b"struct__",
-        *b"typeof__",
-        *b"unsafe__",
+    const KW6: &[[u8; MAX_REPL_LEN]] = &[
+        *b"r#become__",
+        *b"r#extern__",
+        *b"r#return__",
+        *b"r#static__",
+        *b"r#struct__",
+        *b"r#typeof__",
+        *b"r#unsafe__",
     ];
-    const KW7: &[[u8; MAX_KW_LEN]] = &[*b"unsized_", *b"virtual_"];
-    const KW8: &[[u8; MAX_KW_LEN]] = &[*b"abstract", *b"continue", *b"override"];
+    const KW7: &[[u8; MAX_REPL_LEN]] = &[*b"r#unsized_", *b"r#virtual_"];
+    const KW8: &[[u8; MAX_REPL_LEN]] = &[*b"r#abstract", *b"r#continue", *b"r#override"];
 
-    &[KW0, KW1, KW2, KW3, KW4, KW5, KW6, KW7, KW8]
+    &[&[], &[], KW2, KW3, KW4, KW5, KW6, KW7, KW8]
 };
 
-/// Ensure that all strings are UTF-8, because we use `from_utf8_unchecked()` further down.
-#[test]
-fn ensure_utf8() {
-    for kws in KWS {
-        for kw in *kws {
-            if std::str::from_utf8(kw).is_err() {
-                panic!("not UTF-8: {:?}", kw);
-            }
-        }
-    }
-}
+// These ones are only used in the parser, hence why they're private.
+const KWS_EXTRA: &[&[[u8; MAX_REPL_LEN]]] = {
+    const KW4: &[[u8; MAX_REPL_LEN]] = &[*b"r#loop____", *b"r#self____", *b"r#Self____"];
+    const KW5: &[[u8; MAX_REPL_LEN]] = &[*b"r#super___", *b"r#union___"];
+
+    &[&[], &[], &[], &[], KW4, KW5, &[], &[], &[]]
+};
 
 fn is_rust_keyword(ident: &str) -> bool {
+    fn is_rust_keyword_inner(
+        kws: &[&[[u8; MAX_REPL_LEN]]],
+        padded_ident: &[u8; MAX_KW_LEN],
+        ident_len: usize,
+    ) -> bool {
+        // Since the individual buckets are quite short, a linear search is faster than a binary search.
+        kws[ident_len]
+            .iter()
+            .any(|&probe| padded_ident == &probe[2..])
```

Comment:
Interesting! You are right, the

```rust
padded_ident == <[u8; parser::node::MAX_KW_LEN]>::try_from(&probe[2..]).unwrap()
```
"optimization" is not needed anymore, and now the same code generated with and without it. When I first wrote the search, it was still needed because otherwise the assembly would needlessly contain length checks.

Reply:
Oh interesting. I checked and couldn't find out why it was needed. I suppose rustc/LLVM got better.